Add tricky test case blog post (#1941)
theupdateframework/tuf: docs/_posts/testing_a_tricky_edge_case_in_a_tuf_client.md

Timestamp v1:
<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/1/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-03-30T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 1}},
  "spec_version": "1.0.28",
  "version": 1
 }}
</code></pre>

Timestamp v2:

<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/2/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-04-13T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 2}},
  "spec_version": "1.0.28",
  "version": 2
 }}
</code></pre>

## Edge cases are a norm when we talk about security

The TUF spec says the following:
“Record the time at which the update began as the fixed update start time. Time is fixed at the beginning of the update workflow to allow an application using TUF to effectively pause time, in order to ensure that metadata which has a valid expiration time at the beginning of an update does not fail an expiration check later in the update workflow.”

> Now imagine that we have performed an update and stored metadata in a cache. And the locally stored timestamp/snapshot has expired. But we still need it to perform an update from remote by verifying the signatures and we need to use the expired timestamp. Here the repository simulator can do a great job testing this scenario, as we’d like to somehow update expiry without explicitly touching the file system from the tests.

Reviewer: I think this should be moved under `### Example 1:` or line 80.

Author: This is an introduction and is valid for all examples. May be I should remove the `### Example` headers, or rename them. I it just goes under Ex. 1, then Ex. 2 would be missing some context.

Thanks for noticing.

### Example 1:
We can play with versions and expiry to verify that this scenario not explicitly mentioned in the spec works correctly and safely. By using the simulator, we can do the following:

> * Set the timestamp expiry one week ahead (Day 7)
> * On the very first day download, verify, and load metadata for the [top-level roles](https://theupdateframework.github.io/specification/latest/#roles-and-pki) following the specified order and implementing all the checks required in the TUF client workflow (which we simply call a refresh)
> * Then we bump [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) versions to v2 on the same day and set v2 expiry dates three weeks ahead (Day 21)
> 
> Then if we travel in time somewhere between Day 7 and Day 21, we’d like to perform a successful refresh with the expired locally cached timestamp and we want the final version of the snapshot and targets roles to be v2.

Reviewer: I think you can make the last sentence bullets as well and use numbers instead of just bullets.
Then you can add comments referencing the steps.
Also, I think you can split some of the bulltes.
For example, you can do something like this:
```suggestion
1. Set the timestamp expiry one week ahead (day 7).
2. On day 1 download, verify, and load metadata for the [top-level roles](https://theupdateframework.github.io/specification/latest/#roles-and-pki) following the TUF specification order. This is done by simply calling `RespositorySimulator.refresh()`.
3.  On the same day, set [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) expiration date three weeks ahead (day 21).
4. Bump [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) versions to v2 on the same day.
5. Travel in time somewhere between day 7 and day 21.
6. Perform a successful `refresh` (with `RepositorySimulator.refresh()` call).
7. Check that the final version of [snapshot].(https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) is 2.

Keep in mind that when performing the second `refresh` on step 6 [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) has expired, but still, we manage to make a successful refresh to version 2.
```

Author: Thank you! I like your proposal, looks easier to understand.

This is a not so obvious use-case to keep in mind when thinking about updates. You can see how it looks in practice in the [reference implementation](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_metadata).

### Example 2:
Now let’s see if a rollback attack protection can be performed when the local timestamp has expired. In this case we need at least two timestamp and snapshot versions, an expired older version of timestamp, and a verification that a rollback check is performed with the old version.

For a timestamp rollback, the case is pretty similar to the use of expired metadata. We can do the following:
* Set timestamp v1 expiry one week ahead (Day 7)
* Perform refresh on the very first day
* Publish timestamp v2 with expiry three weeks ahead (Day 21)
* Perform a refresh somewhere between Day 7 and Day 21

> All we need is to verify that rollback check uses the expired timestamp v1. (For reference, see the implementation [example](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_timestamp_version_rollback)).

Reviewer: I think it will be better if you move this sentence to the bullet points.
With the usage of bullet points (or numbers), one will expect that all of the test steps are described, but here they are not. There is one additional step to check which timestamp is used which is mentioned here.

---
title: "Testing Tricky Edge Cases in a TUF Client"
author: Ivana Atanasova
---

Usually the TUF Specification creates an impression of simple and straightforward approach to address software update systems security gaps. In the next few paragraphs we'll try to convince you that the devil is in the details.

With the [v1.0.0 release](https://blogs.vmware.com/opensource/2022/02/22/python-tuf-reaches-version-1-0-0/) we can say that the current reference implementation is finally in a good place, although it wouldn’t be so trustworthy without all the awesome test functionality it provides. Therein lies some interesting surprises, for the test coverage reflects use cases that wouldn’t easily come to mind. TUF, in fact, is capable of managing some tricky business! And that’s why it provides more than just test coverage - it is a way to analyze and validate that the implementation strictly follows the TUF spec with all its corner cases and tricky details that are far from obvious.

> Before looking into them, we need to introduce the test functionality itself.

Reviewer 2: I still don't fully understand what you are trying to say here (see https://github.com/theupdateframework/python-tuf/pull/1941#discussion_r871088189).

## Some repository simulator magic

The test suite is heavily based on [RepositorySimulator](https://github.com/theupdateframework/python-tuf/blob/develop/tests/repository_simulator.py), which allows you to play with repository metadata by modifying it, signing and storing new roles versions, while serving older ones in the client test code. You can also simulate downloading new metadata from a remote without the need of file access or network connections, and modify expiry dates and time.

Even though `RepositorySimulator` hosts repos purely in memory, you can supply the `--dump` flag to write its contents to a temporary directory on the local filesystem with "/metadata/..." and "/targets/..." URL paths that host metadata and targets respectively in order to audit the metadata. The test suite provides you with the ability to see the "live" test repository state for debugging purposes.

Let’s cite a specific example with testing expired metadata to demonstrate the cool thing the `RepositorySimulator` provides, i.e. the capability to simulate real repository chains of updates as suggested by the spec, and not just modify individual metadata.

> More specifically, we would like to simulate a workflow in which a [targets](https://theupdateframework.github.io/specification/latest/#targets) version is being increased and a [timestamp](https://theupdateframework.github.io/specification/latest/#timestamp) expiry date is being changed. We are going to elaborate in more detail in the sections below about an actual use-case this can be used to test the `Updater` above all programmatically. Now, let's just focus on how to verify that the `RepositorySimulator` did what we expected.

Reviewer 2: Minor wording nit: "...elaborate in more detail in the sections below about an actual use-case this can be used to test..." reads a bit convoluted. What about "...elaborate below on how this can be used to test ..."?

Let's assume we did the following:
* Upgraded `targets` to v2
* Changed `timestamp` v2 expiry date

We can verify that the metadata looks as expected, without the need to implement file access.

First, we need to find the corresponding temporary directory:
```
$ python3 test_updater_top_level_update.py TestRefresh.test_expired_metadata --dump
Repository Simulator dumps in /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_
```

Once we know it, we can verify that the metadata has 2 cached versions:

```
$ ls /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/
1	2
```
```
$ ls /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/1
1.root.json	snapshot.json	targets.json	timestamp.json

> ```

Reviewer 2: Just an idea: Maybe it would be worth to `tree /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata`, instead of `ls`ting it twice?

And now we can also see that after bumping the version and moving timestamp v2 expiry date two weeks forward from v1, the v2 corresponding timestamp metadata has recorded that expiry date correctly:

Timestamp v1:
<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/1/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-03-30T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 1}},
  "spec_version": "1.0.28",
  "version": 1
 }}
</code></pre>

Timestamp v2:

<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/2/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-04-13T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 2}},
  "spec_version": "1.0.28",
  "version": 2
 }}
</code></pre>

As you can see, the first date is 30 Mar and the second - 13 Apr, which is exactly 14 days later. This is a great way to observe what the tests really do and check if they do it successfully.

## When we talk about security, edge cases are the norm

> Now, let’s take a closer look at two edge cases, using in this test the two cool things the RepositorySimulator provides:

Reviewer 2: What are "the two cool things the RepoSimulator provides"? Also, further above you say "demonstrate the cool thing the RepositorySimulator provides". Does it provide one or two cool things now? :P

### Example with expired metadata:

> Imagine that we have performed an update and stored metadata in a cache. And the locally stored timestamp/snapshot has expired. But we still need it to perform an update from remote by verifying the signatures and we need to use the expired timestamp. Here the repository simulator can do a great job testing this scenario, as we’d like to somehow update expiry without explicitly touching the file system from the tests.

Reviewer 2: Why is it important to not touch the file system?

We can play with versions and expiry to verify that this scenario not explicitly mentioned in the spec works correctly and safely. By using the simulator, we can do the following:
1. Set the timestamp expiry one week ahead (to day 7)
2. On the very first day (day 0) download, verify, and load metadata for the [top-level roles](https://theupdateframework.github.io/specification/latest/#roles-and-pki) following the TUF specification order. This is done by simply calling `updater.refresh()`.
3. Then we bump [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) versions to v2 in the repository on the same day (day 0)
4. Set v2 expiry dates three weeks ahead (to day 21)
5. Travel in time somewhere between day 7 and day 21
6. Perform a successful `refresh` (with `updater.refresh()` call) with the expired locally cached timestamp

> 7. Check that the final repository version of the snapshot and targets roles is v2.

Reviewer 2: Minor format nit: I think you want a blank line before the next sentence, otherwise it looks like the summary is part of step 7.

---
title: "Testing Tricky Edge Cases in a TUF Client"
author: Ivana Atanasova
---

Usually the TUF Specification creates an impression of simple and straightforward approach to address software update systems security gaps. In the next few paragraphs we'll try to convince you that the devil is in the details.

> With the [v1.0.0 release](https://blogs.vmware.com/opensource/2022/02/22/python-tuf-reaches-version-1-0-0/) we can say that the current reference implementation is finally in a good place, although it wouldn’t be so trustworthy without all the awesome test functionality it provides. Therein lies some interesting surprises, for the test coverage reflects use cases that wouldn’t respectively come to mind. TUF, in fact, is capable of managing some tricky business! And that’s why it provides more than just test coverage - it is a way to analyze and validate that the implementation strictly follows the TUF spec with all its corner cases and tricky details that are far from obvious.

Reviewer 2: I am unsure if the word "respectively" fits here (cc @jhdalek55).

Reviewer 2: Could you try to rephrase:
> TUF, in fact, is capable of managing some tricky business! And that’s why it provides more than just test coverage - it is a way to analyze and validate that the implementation strictly follows the TUF spec with all its corner cases and tricky details that are far from obvious.

 "managing some tricky business" reads a bit overly casual and vague. Also, I don't understand why the consequence is that "it provides more than just test coverage". What is "it" and what do you mean by "provide test coverage"?

Before looking into them, we need to introduce the test functionality itself.

## Some Repository Simulator Magic

> 

Reviewer 2: I wouldn't start this sentence with "The great thing about RepositorySimulator is ...". The reader doesn't know yet what RepositorySimulator is and why they should care about it. Maybe you can set the stage for the tool before you praise it. Something along the lines of "The test suite is heavily based on RepositorySimulator, which allows you to ..."

> The great thing about [RepositorySimulator](https://github.com/theupdateframework/python-tuf/blob/develop/tests/repository_simulator.py), is it allows you to play with repository metadata by modifying it, signing and storing new roles versions, while serving older ones in the client test code. You can also simulate downloading new metadata from a remote without the need of file access or network connections, and modify expiry dates and time.

Reviewer 2: Premature comma: 
```suggestion
The great thing about [RepositorySimulator](https://github.com/theupdateframework/python-tuf/blob/develop/tests/repository_simulator.py) is, it allows you to play with repository metadata by modifying it, signing and storing new roles versions, while serving older ones in the client test code. You can also simulate downloading new metadata from a remote without the need of file access or network connections, and modify expiry dates and time.
```

> When you execute a test in python-tuf, it creates a local temporary directory serving for cache, with "/metadata/..." and "/targets/..." URL paths that host metadata and targets respectively.

Reviewer 2: When you execute a test ... **with repository simulator**?

Also, "cache" seems a bit like an understatement. RepositorySimulator simulates a TUF repository (for metadata and target files), right?

Reviewer 3: oh yeah this is actually incorrect: RepositorySimulator does not create directories, it serves everything from memory

Author: so probably it should stay just "in python-tuf"?

Reviewer 2: This could actually make for a good transition to advertising the `--dump` flag. Something along the lines of, "Although, RepositorySimulator hosts repos purely in memory, you can supply the `--dump` flag to write its contents to a temporary directory on the local filesystem in order to audit the metadata."  (feel free to rephrase as you see fit)

> The test suite provides you with the ability to see the "live" test repository state for debugging purposes. You just need to use the `--dump` option, so that you can see the current tests corresponding path.

Reviewer 2: I'd remove the second sentence here. It's a usage detail, which you explain below anyway.

Let’s cite a specific example with testing expired metadata to demonstrate two of the cool things the RepositorySimulator provides:
* Upgrade [targets](https://theupdateframework.github.io/specification/latest/#targets) to v2

> * Change [timestamp](https://theupdateframework.github.io/specification/latest/#timestamp) v2 expiry date

Reviewer 2: I think the cool thing here is the bigger picture, i.e. the whole repository update story `test_expired_metadata` tells. 

In other words, I would try to emphasize on RepositorySimulator's capability to simulate real repository chains of updates as suggested by the spec, and not just modify individual metadata.

> So how can we verify that the metadata looks as expected, without the need to implement file access?

Reviewer 2: I feel like you are putting a lot of emphasis on the `--dump` feature, which is cool but not the core of the tool. I'd talk more about the specific workflow that RepositorySimulator simulates here and how that can be used to test the Updater above all programmatically, and only then mention how we can pass the handy `--dump` flag to audit what the simulator is doing.

1.root.json	snapshot.json	targets.json	timestamp.json
```

And now we can also see that after bumping the version and moving timestamp v2 expiry date two weeks forward from v1, the v2 corresponding timestamp metadata has recorded that expiry date correctly:

Timestamp v1:
<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/1/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-03-30T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 1}},
  "spec_version": "1.0.28",
  "version": 1
 }}
</code></pre>

Timestamp v2:

<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/2/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-04-13T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 2}},
  "spec_version": "1.0.28",
  "version": 2
 }}
</code></pre>

As you can see, the first date is 30 Mar and the second - 13 Apr, which is exactly 14 days later. This is a great way to observe what the tests really do and check if they do it successfully.

> ## When We Talk about Security, Edge Cases are the Norm

Reviewer 2: Minor nit: Please choose a consistent [capitalization style](https://en.wikipedia.org/wiki/Capitalization#Names_of_capitalization_styles) (also review other titles)

Now, let’s take a closer look at two edge cases, using in this test the two cool things the RepositorySimulator provides:

> The TUF spec says the following:

Reviewer 2: Can you add a bit of context, when/where/why the TUF spec says this?

Reviewer 2: Having read the examples below, I don't understand why we need this paragraph from the spec here.

“Record the time at which the update began as the fixed update start time. Time is fixed at the beginning of the update workflow to allow an application using TUF to effectively pause time, in order to ensure that metadata which has a valid expiration time at the beginning of an update does not fail an expiration check later in the update workflow.” Let's see the following examples where this is not so obvious.

> Let's consider the following examples where this is not so obvious.

Reviewer 2: Duplicate sentence.

### Example with expired metadata:

Imagine that we have performed an update and stored metadata in a cache. And the locally stored timestamp/snapshot has expired. But we still need it to perform an update from remote by verifying the signatures and we need to use the expired timestamp. Here the repository simulator can do a great job testing this scenario, as we’d like to somehow update expiry without explicitly touching the file system from the tests.

We can play with versions and expiry to verify that this scenario not explicitly mentioned in the spec works correctly and safely. By using the simulator, we can do the following:
1. Set the timestamp expiry one week ahead (Day 7)

> 2. On the very first day download, verify, and load metadata for the [top-level roles](https://theupdateframework.github.io/specification/latest/#roles-and-pki) following the TUF specification order. This is done by simply calling `RespositorySimulator.refresh()`.

Reviewer 2: You mean `updater.refresh`, right? (same below)

3. Then we bump [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) versions to v2 on the same day
4. Set v2 expiry dates three weeks ahead (Day 21)
5. Travel in time somewhere between Day 7 and Day 21
6. Perform a successful `refresh` (with `RepositorySimulator.refresh()` call) with the expired locally cached timestamp
7. Check that the final version of the snapshot and targets roles is v2.

> This is a not so obvious use-case to keep in mind when thinking about updates. You can see how it looks in practice in the [reference implementation](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_metadata).

Reviewer 2: I must say, the docstring from `test_expired_metadata` describes this chain of events slightly clearer:
- it mentions what happens in the repo and what happens on the client
- it mentions for each event when it happens in a uniform way

Reviewer 2: (same applies in example below)

Author: Do you think it's better to cite the docstring, or link to it, or cite and elaborate?

Reviewer 2: Cite and elaborate probably. :) I just wanted to point out that currently it is not immediately obvious if something happens in the repo or on the client. If you polish that a bit, I'm happy. My second comment about the time is less important.

For a timestamp rollback, the case is pretty similar to the use of expired metadata. We can do the following:
1. Set timestamp v1 expiry one week ahead (Day 7)
2. Perform refresh on the very first day
3. Publish timestamp v2 with expiry three weeks ahead (Day 21)
4. Perform a refresh somewhere between Day 7 and Day 21
5. Verify that rollback check uses the expired timestamp v1. (For reference, see the implementation [example](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_timestamp_version_rollback)).

A similar approach can be used when testing both timestamp and snapshot rollback protection. We just need to guarantee that after the last snapshot update, the snapshot version is not the latest in order to verify a rollback check is performed both with expired timestamp and an older snapshot. Sounds complicated, but it’s pretty easy with the simulator and [this example](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_timestamp_snapshot_rollback) illustrates it pretty well. 

> 

Reviewer 2: Are these two examples really an edge cases? Aren't they (vaguely) defined by the client workflow in the spec?

I do agree that they are tricky to implement in a client and tricky to test, and that the RepositorySimulator does a great job in achieving the latter.

Reviewer 3: I probably described it as an edge case :grimacing: with the logic that 
* we want to ensure that we only use valid metadata
* expired metadata is not valid
* **except** when were validating new metadata with an older metadata (in this case specifically old metadata can be expired but must be otherwise valid)

this is defined in spec, that's true

## The Devil is in the Details

> One of the great things about a reference implementation is that one can learn a lot about the TUF specification by looking at the tests, which are full of examples that would hardly come to mind when you read the illusory straightforward workflow explained in the spec. And those tests most likely do not cover everything… 

Reviewer 2: Maybe say "abstract" instead of "illusory"?

---
title: "Testing a tricky edge case in a TUF client"
author: Ivana Atanasova
---

> Usually the TUF Specification creates an impression of simple and straightforward approach to address software update systems security gaps. In the next few paragraphs we'll try to convince you that the devil is in the details.

Reviewer: I am intrigued in hearing from the rest of the team if they think the specification is simple and straightforward. :D
Maybe it's not bad in using a little advertainment language.

Author: The key is in "creates an impression" :D

Timestamp v1:
<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/1/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-03-30T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 1}},
  "spec_version": "1.0.28",
  "version": 1
 }}
</code></pre>

Timestamp v2:

<pre><code>$ cat /var/folders/pr/b0xyysh907s7mvs3wxv7vvb80000gp/T/tmpzvr5xah_/test_expired_metadata/2/timestamp.json 
{
 "signatures": [{...}],
 "signed": {
  "_type": "timestamp",
  <b>"expires": "2022-04-13T00:18:31Z"</b>,
  "meta": { "snapshot.json": {"version": 2}},
  "spec_version": "1.0.28",
  "version": 2
 }}
</code></pre>

## Edge cases are a norm when we talk about security

The TUF spec says the following:
“Record the time at which the update began as the fixed update start time. Time is fixed at the beginning of the update workflow to allow an application using TUF to effectively pause time, in order to ensure that metadata which has a valid expiration time at the beginning of an update does not fail an expiration check later in the update workflow.”

> Now imagine that we have performed an update and stored metadata in cache. And the locally stored timestamp/snapshot has expired. But we still need it to perform an update from remote by verifying the signatures and we need to use the expired timestamp. Here the repository simulator can do a great job, as we’d like to somehow update expiry without explicitly touching the file system from the tests.

Reviewer: ```suggestion
Now imagine that we have performed an update and stored metadata in a cache. And the locally stored timestamp/snapshot has expired. But we still need it to perform an update from remote by verifying the signatures and we need to use the expired timestamp. Here the repository simulator can do a great job testing this scenario, as we’d like to somehow update expiry without explicitly touching the file system from the tests.
```
I think it should be underlined clearer that the `RepositorySimulator` can be useful for testing this scenerio, not actually performing it.

> We can play with versions and expiry to verify that this scenario not explicitly mentioned in the spec works correctly and safely. By using the simulator, we can set the timestamp expiry one week ahead. Then on the very first day we download, verify, and load metadata for the [top-level roles](https://theupdateframework.github.io/specification/latest/#roles-and-pki) following the specified order and implementing all the checks required in the TUF client workflow (which we simply call a refresh). Then we bump [snapshot](https://theupdateframework.github.io/specification/latest/#update-snapshot) and [targets](https://theupdateframework.github.io/specification/latest/#targets) versions to v2 on the same day and set v2 expiry dates three weeks ahead. If we travel in time somewhere between day 7 and day 21, we’d like to perform a successful refresh with the expired locally cached timestamp and the final version of the snapshot and targets roles to be v2.

Reviewer: Can you enumerate the steps you are describing here and split them one by one, so they could be easier to follow?
It's a little hard for me to follow.

This is a not so obvious use-case to keep in mind when thinking about updates. You can see how it looks in practice in the [reference implementation](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_metadata).

Now let’s see if a rollback attack protection can be performed when the local timestamp has expired. In this case we need at least two timestamp and snapshot versions, an expired older version of timestamp, and a verification that a rollback check is performed with the old version.

> For a timestamp rollback, the case is pretty similar to the use of  expired metadata. We can set timestamp v1 expiry one week ahead, perform refresh on the very first day, publish timestamp v2 with expiry three weeks ahead and perform a refresh somewhere between day 7 and day 21. All we need is to verify that rollback check uses the expired timestamp v1. (For reference, see the implementation [example](https://github.com/theupdateframework/python-tuf/blob/develop/tests/test_updater_top_level_update.py#:~:text=test_expired_timestamp_version_rollback)).

Reviewer: Maybe you add a new space to actually divide the two paragraphs because markdown merges them as if there is no newline between lines number 83 and 84?

Reviewer: Again as above, I recommend using numbers and writing them in separate steps.
